components/supervisor/pkg/ports/ports-config.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// GetFromRange returns the port config from the port range config
+func (configs *Configs) GetFromRange(port uint32) (*gitpod.PortConfig, bool) {
+	if configs == nil {
+		return nil, false
+	}
+	for _, rangeConfig := range configs.instanceRangeConfigs {
+		if rangeConfig.Start <= port && port <= rangeConfig.End {
+			return &gitpod.PortConfig{
+				Port:       float64(port),
+				OnOpen:     rangeConfig.OnOpen,
+				Visibility: rangeConfig.Visibility,
+			}, true
+		}
+	}
+	return nil, false
+}
+
+// ConfigInterace allows to watch port configurations
+type ConfigInterace interface {
+	// Observe provides channels triggered whenever the port configurations are changed.
+	Observe(ctx context.Context) (<-chan *Configs, <-chan error)
+}
+
+// ConfigService allows to watch port configurations
+type ConfigService struct {
+	workspaceID   string
+	configService gitpod.ConfigInterface
+	gitpodAPI     gitpod.APIInterface
+	parser        *configParser
+}
+
+// NewConfigService creates a new instance of ConfigService
+func NewConfigService(workspaceID string, configService gitpod.ConfigInterface, gitpodAPI gitpod.APIInterface) *ConfigService {
+	return &ConfigService{
+		workspaceID:   workspaceID,
+		configService: configService,
+		gitpodAPI:     gitpodAPI,
+		parser:        newConfigParser(),
+	}
+}
+
+// Observe provides channels triggered whenever the port configurations are changed.
+func (service *ConfigService) Observe(ctx context.Context) (<-chan *Configs, <-chan error) {
+	updatesChan := make(chan *Configs)
+	errorsChan := make(chan error, 1)
+
+	go func() {
+		defer close(updatesChan)
+		defer close(errorsChan)
+
+		configs, errs := service.configService.Observe(ctx)
+
+		current := &Configs{}
+		if service.gitpodAPI != nil {
+			info, err := service.gitpodAPI.GetWorkspace(ctx, service.workspaceID)
+			if err != nil {
+				errorsChan <- err
+			} else {
+				current.workspaceConfigs = service.parser.parseWorkspaceConfigs(info.Workspace.Config.Ports)
+				updatesChan <- &Configs{workspaceConfigs: current.workspaceConfigs}
+			}
+		} else {
+			errorsChan <- errors.New("could not connect to Gitpod API to fetch workspace port configs")
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-errs:
+				errorsChan <- err
+			case config := <-configs:
+				if service.update(config, current) {
+					updatesChan <- &Configs{
+						workspaceConfigs:     current.workspaceConfigs,
+						instancePortConfigs:  current.instancePortConfigs,
+						instanceRangeConfigs: current.instanceRangeConfigs,
+					}
+				}
+			}
+		}
+	}()
+	return updatesChan, errorsChan
+}
+
+func (service *ConfigService) update(config *gitpod.GitpodConfig, current *Configs) bool {
+	currentPortConfigs, currentRangeConfigs := current.instancePortConfigs, current.instanceRangeConfigs
+	var ports []*gitpod.PortsItems
+	if config != nil {
+		ports = config.Ports
+	}
+	portConfigs, rangeConfigs := service.parser.parseInstanceConfigs(ports)
```

Comment:
It took me a minute to understand this function. If I understand correctly, when there's no config, `service.parser.parseInstanceConfigs(ports)` will always return empty lists. In that case we could exit early returning `false`.

Reply:
i don't think it would be correct, if the file was removed we should remove the previous state as well